### PR TITLE
Introduce `Stateful` PrecisionPlugin

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

part of #11429: introduce `Stateful` PrecisionPlugin (see Appendix code changes outlined in issue)

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -192,7 +192,17 @@ def restore_training_state(self) -> None:
...
 
         # restore precision plugin (scaler etc.)
-        self.trainer.precision_plugin.on_load_checkpoint(self._loaded_checkpoint)
+        prec_plugin = self.trainer.precision_plugin
+        prec_plugin.on_load_checkpoint(self._loaded_checkpoint)
+        if prec_plugin.__class__.__name__ in self._loaded_checkpoint:
+            prec_plugin.load_state_dict(self._loaded_checkpoint[prec_plugin.__class__.__name__])
+
+        # old checkpoints compatibility
+        # should we raise error and force user to run utilities/upgrade_checkpoint instead?
```
Comment (jjenniferdai) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:201]: see comment

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -389,6 +401,8 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
         model.on_save_checkpoint(checkpoint)
         if self.trainer.datamodule is not None:
             self.trainer.datamodule.on_save_checkpoint(checkpoint)
+        if not weights_only:
+            self.trainer.precision_plugin.on_save_checkpoint(checkpoint)
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:405]: why does this happen after the other on_save_checkpoint hooks? let's keep the same order as before

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:201]: cc @awaelchli - we must keep backwards compatibility with old checkpoints. if we can do this best effort with comments about including new keys & falling back to old keys for precision-related logic, we should do that.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -192,7 +192,17 @@ def restore_training_state(self) -> None:
...
+        prec_plugin.on_load_checkpoint(self._loaded_checkpoint)
+        if prec_plugin.__class__.__name__ in self._loaded_checkpoint:
+            prec_plugin.load_state_dict(self._loaded_checkpoint[prec_plugin.__class__.__name__])
+
+        # old checkpoints compatibility
+        # should we raise error and force user to run utilities/upgrade_checkpoint instead?
+        if "amp_scaling_state" in self._loaded_checkpoint:
+            prec_plugin.load_state_dict(self._loaded_checkpoint["amp_scaling_state"])
+        if "native_amp_scaling_state" in self._loaded_checkpoint:
+            prec_plugin.load_state_dict(self._loaded_checkpoint["native_amp_scaling_state"])
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:205]: this is now sufficiently complicated enough that it can be its own function within this class

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 4]
```diff
@@ -372,7 +382,9 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
                 lr_schedulers.append(config.scheduler.state_dict())
             checkpoint["lr_schedulers"] = lr_schedulers
 
-            self.trainer.precision_plugin.on_save_checkpoint(checkpoint)
+            # precision plugin
+            prec_plugin = self.trainer.precision_plugin
+            checkpoint[prec_plugin.__class__.__name__] = self.trainer.precision_plugin.state_dict()
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:387]: note: now if the precision plugin/module ever changes, loading old checkpoints will fail

**Review by ananthsub — COMMENTED** [review]
as a follow up, please deprecate CheckpointHooks off the PrecisionPlugin interface: https://github.com/PyTorchLightning/pytorch-lightning/blob/86b177ebe5427725b35fde1a8808a7b59b8a277a/pytorch_lightning/plugins/precision/precision_plugin.py#L29

It is no longer needed following the Stateful change and should be removed in v1.8

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 4 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:387]: https://github.com/PyTorchLightning/pytorch-lightning/blob/86b177ebe5427725b35fde1a8808a7b59b8a277a/pytorch_lightning/callbacks/base.py#L43 

let's use qualname here too for consistency

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:201]: Making users upgrade checkpoints is tedious. Let's avoid it for as long as possible.

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:201]: >  should we raise error and force user to run utilities/upgrade_checkpoint instead?

Neither IMO. Lightning should upgrade the checkpoint directly when loading the file, and before sending the dict to the internal components. The development of this feature is currently stale: #9396
Here for example, an upgrade would consist of detecting the keys amp_scaling_state, native_amp_scaling_state, and merging them into the new format saved under the new key `NativeMixedPrecisionPlugin.__class.__name__` or whatever it ends up being :)

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 5]
```diff
@@ -92,9 +92,9 @@ def optimizer_step(
...
-    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
-        if "amp_scaling_state" in checkpoint:
-            amp.load_state_dict(checkpoint["amp_scaling_state"])
+    def state_dict(self) -> Dict[str, Any]:
+        return amp.state_dict()
 
-    def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
-        checkpoint["amp_scaling_state"] = amp.state_dict()
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        if state_dict:
```
Comment (tchaton) [inline pytorch_lightning/plugins/precision/apex_amp.py:99]: Is this required ? Would it fail if empty ?

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 6]
```diff
@@ -106,10 +106,10 @@ def forward_context(self) -> Generator[None, None, None]:
...
 
-    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
-        if self.scaler is not None and "native_amp_scaling_state" in checkpoint:
-            self.scaler.load_state_dict(checkpoint["native_amp_scaling_state"])
-
-    def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
+    def state_dict(self) -> Dict[str, Any]:
         if self.scaler is not None:
-            checkpoint["native_amp_scaling_state"] = self.scaler.state_dict()
+            return self.scaler.state_dict()
```
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/native_amp.py:111]: this returns None if scaler isn't set. should this return an empty dict instead? if so, does the checkpoint connector filter out empty dicts from being saved or loaded?

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 7]
```diff
@@ -109,7 +109,9 @@ def forward_context(self) -> Generator[None, None, None]:
     def state_dict(self) -> Dict[str, Any]:
         if self.scaler is not None:
             return self.scaler.state_dict()
+        else:
+            return {}
```
Comment (ananthsub) [inline pytorch_lightning/plugins/precision/native_amp.py]:
```suggestion
        if self.scaler is not None:
            return self.scaler.state_dict()
        return {}
```

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 5 above)
Comment (jjenniferdai) [reply pytorch_lightning/plugins/precision/apex_amp.py]: (moved the empty check to just not save if empty)

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:201]: I agree. For now, we can keep it as it is and refactor with #9396.

```suggestion
        # old checkpoints compatibility
```

File: pytorch_lightning/plugins/precision/apex_amp.py [hunk 8]
```diff
@@ -92,9 +92,8 @@ def optimizer_step(
         if not isinstance(model, pl.LightningModule) or not model.automatic_optimization or not skipped_backward:
             optimizer.step(**kwargs)
 
-    def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/apex_amp.py:95]: This will be a breaking change if a subclass is calling the super implementation, consider keeping an empty shell during deprecation

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 9]
```diff
@@ -242,3 +242,20 @@ def teardown(self) -> None:
 
         It is the right place to release memory and free other resources.
         """
+
+    def state_dict(self) -> Dict[str, Any]:
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/precision_plugin.py:246]: Did you consider making the class use `Stateful` as a mixin?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 10]
```diff
@@ -192,7 +192,17 @@ def restore_training_state(self) -> None:
             return
 
         # restore precision plugin (scaler etc.)
-        self.trainer.precision_plugin.on_load_checkpoint(self._loaded_checkpoint)
+        prec_plugin = self.trainer.precision_plugin
+        prec_plugin.on_load_checkpoint(self._loaded_checkpoint)
+        if prec_plugin.__class__.__qualname__ in self._loaded_checkpoint:
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:197]: Here you do `on_load_checkpoint` -> `load_state_dict`

but below: `state_dict` -> `on_save_checkpoint`.

Is this on purpose?

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 8 above)
Comment (justusschock) [reply pytorch_lightning/plugins/precision/apex_amp.py:95]: I think it shouldn't be an empty-shell, but it should raise a warning and call either `state_dict` or `load_state_dict` for convenience. Usually that's what people want to do here :)

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 10 above)
Comment (justusschock) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:197]: I would always have the old api falling back to the new one with a warning and *never* the other way around :)

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 9 above)
Comment (Borda) [reply pytorch_lightning/plugins/precision/precision_plugin.py:246]: just was thinking the same as I see it already a few times...

**Review by Borda — APPROVED** [review]
@carmocca let's merge this as it is and create an issue bout unifying the stateful mixin?

**Comment by carmocca** [comment]
> let's merge this as it is and create an issue bout unifying the stateful mixin?

We can wait for @jjenniferdai to address comments.

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 4 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: > note: now if the precision plugin/module ever changes, loading old checkpoints will fail

if I understand this comment correctly - for compatibility with old checkpoint keys, added an `isinstance` check for the precision plugin before calling `load_state_dict(ckpt["old_precplugin_key"])`

File: pytorch_lightning/plugins/precision/apex_amp.py (hunk 8 above)
Comment (jjenniferdai) [reply pytorch_lightning/plugins/precision/apex_amp.py:95]: good point, I think an empty shell will do bc lightning takes care of saving/loading states in `state_dict/load_state_dict` now.

The only need to keep the actual old `on_save/load_checkpoint` contents is if users manually check `ckpt["old_precplugin_key"]` somewhere in their code... do we need a deprecation process to support this? e.g. write precision plugin state_dict under both `"old_precplugin_key"` and under `qualname`

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 10 above)
Comment (jjenniferdai) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: save: `state_dict` needs to come before `on_save_checkpoint` in order for `on_save_checkpoint` to receive the "full" ckpt dict argument.

load: not as definitive as save, but I was thinking `on_load_checkpoint` gets the loaded ckpt dict to do any necessary logic before passing it to `load_state_dict` calls - thoughts?

File: pytorch_lightning/plugins/precision/precision_plugin.py (hunk 9 above)
Comment (jjenniferdai) [reply pytorch_lightning/plugins/precision/precision_plugin.py:246]: originally proposed `Protocol` so that users could simply implement `state_dict/load_state_dict` without having to extend anything. but now that we're adding them to lightning base classes, I think mixin could be better - can change this in another PR

File: pytorch_lightning/plugins/precision/precision_plugin.py [hunk 11]
```diff
@@ -242,3 +242,20 @@ def teardown(self) -> None:
...
+        """Called when saving a checkpoint, implement to generate and save precision plugin state.
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/precision/precision_plugin.py:247]:
```suggestion
        """Called when saving a checkpoint, implement to generate precision plugin state_dict.
```
the hook specifically doesn't save it.

**Review by rohitgr7 — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 10 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: @jjenniferdai please add that as a comment directly in the code

File: pytorch_lightning/plugins/precision/native_amp.py [hunk 12]
```diff
@@ -106,10 +107,23 @@ def forward_context(self) -> Generator[None, None, None]:
...
+        return {}
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        if self.scaler is not None:
+            self.scaler.load_state_dict(state_dict)
+
     def on_load_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
-        if self.scaler is not None and "native_amp_scaling_state" in checkpoint:
-            self.scaler.load_state_dict(checkpoint["native_amp_scaling_state"])
+        rank_zero_deprecation(
```
Comment (carmocca) [inline pytorch_lightning/plugins/precision/native_amp.py:120]: Missing deprecation tests for these 2 here and in the apex plugin

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/plugins/precision/native_amp.py (hunk 12 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/precision/native_amp.py:122]: so these are  hooks called by Lightning, not the user. Only if the user overrides these hooks AND calls super will they every see this warning, which is less likely. 

I would put the warning into the docstring instead. 
Optionally one can use the is_overridden utility in the place where this hook gets called.

Comment (jjenniferdai) [reply pytorch_lightning/plugins/precision/native_amp.py:122]: - protecting against `super.on_save/load_checkpoint()` calls: presumably this would only happen inside an overriden `on_save/load_checkpoint`. 
- protecting against overriding `on_save/load_checkpoint()`: this itself actually does not need any warnings - until issue #11824

we can either do
1. `if is_overridden: deprecation("if you are calling super.on_save/load_checkpoint, super method is deprecated.)`, then after issue #11824 the warning message just becomes `"checkpointhooks are deprecated"`
2. or we can just do docstring for now until  `"checkpointhooks are deprecated"` entirely

I'm thinking 2

Comment (ananthsub) [reply pytorch_lightning/plugins/precision/native_amp.py:122]: I think docstring for now and deprecation in a follow up is reasonable

Comment (awaelchli) [reply pytorch_lightning/plugins/precision/native_amp.py:122]: yes

**Review by awaelchli — APPROVED** [review]
Unfortunately we never had any tests around amp checkpointing (I couldn't find any?). I created this issue #11885 because we need some form of checks to be confident with refactors like these.

**Review by Borda — APPROVED** [review]
lgtm
